Support int32 indices and offsets in nn.EmbeddingBag

qizzzh opened the pull request:
Summary:
It's in general helpful to support int32 indices and offsets, especially when such tensors are large and need to be transferred to accelerator backends. Since it may not be very useful to support the combination of int32 indices and int64 offsets, here we enforce that these two must have the same type.

Tensor indexing utilities all assume indices should be int64. I'd like to address that in a follow up diff as this one is already pretty big.

Test Plan: unit tests

Differential Revision: D24470808

ngimel commented on test/test_nn.py:
```diff
@@ -11928,55 +11936,60 @@ def _test_EmbeddingBag(self, device, mode, sparse, dtype=torch.double, test_back
+            self._test_EmbeddingBag(device, 'sum', True, wdtype, dtype, test_backward=test_backward)
+            self._test_EmbeddingBag(device, 'mean', True, wdtype, dtype, test_backward=test_backward)
 
     @dtypesIfCUDA(torch.half, torch.float, torch.double)
```
this does not look right, on cuda dtype will be half, float and double instead of integral. Same comment as above for specifying tuples of dtypes

qizzzh replied:
Thanks for the catch. Will fix.

ngimel reviewed (commented):
Thank you for working on this, this is a commonly needed feature. 
This generally looks good, can you please update the documentation to indicate that indices/offsets can now be int, and that they have to be of the same type?

qizzzh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -32,11 +32,17 @@ namespace native {
-  offset2bag = offset2bag.cumsum(0);     // offset2bag = [0 0 1 1 2]
+static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
+  auto ones = at::ones_like(offsets, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  AT_DISPATCH_INDEX_TYPES(offsets.scalar_type(), "make_offset2bag", [&] () {
```
> Don't want to convert offsets to long

Yes. Would you suggest to keep using the original code with offsets.long() or the current one with literal ones? Seems the later may have better perf.

qizzzh commented on aten/src/ATen/native/Embedding.cpp:
```diff
@@ -68,65 +68,68 @@ Tensor embedding_sparse_backward(
+    if (numel > 1000) {
+      // The strategy is to parallelize over sections of the vocabulary, so that
+      // thread 1 handles updates to gradWeight[0..nVocab/nThreads]. Every thread
+      // has to traverse the entire input, but the dominating factor is the axpy
```
Good catch. Let me remove it.

qizzzh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -172,27 +180,28 @@ void index_select_add<float>(const Tensor &select_indices,
+  auto* scale_data = scale.data_ptr<data_t>();
   auto scale_stride = scale.stride(0);
 
   for (int64_t i = 0; i < numel; i++) {
```
Not sure about the history..

qizzzh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -598,23 +615,24 @@ Tensor _embedding_bag_backward(const Tensor &grad, const Tensor &indices,
   if (indices.numel() != 0 && offset2bag.numel() == 0) {
     offset2bag_ = at::zeros(
-       {indices.sizes()[0] + 1}, indices.options()); // offset2bag = [0 0 0 0 0]
+       {indices.sizes()[0] + 1}, offsets.options()); // offset2bag = [0 0 0 0 0]
```
I think `offset2bag_` is closer to `offsets` - dims, type

qizzzh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -638,21 +656,32 @@ static Tensor _embedding_bag_dense_backward_cpu_max(
-    index_grad_weight.select(1, dim).index_add_(
-      0, nonempty_max_indices.select(1, dim), nonempty_grad.select(1, dim));
+    //index_grad_weight.select(1, dim).index_add_(
+    //  0, nonempty_max_indices.select(1, dim), nonempty_grad.select(1, dim));
```
Will add.

qizzzh commented on test/test_nn.py:
```diff
@@ -11593,16 +11596,17 @@ def _embedding_bag_reference_impl(self, input, weight, offsets=None, mode='sum',
         return torch.stack(bags)
 
-    def test_EmbeddingBag_empty_per_sample_weights_and_offsets(self, device):
+    @dtypes(torch.int, torch.long)
```
Thanks for the suggestion. Will update.

dzhulgakov commented on aten/src/ATen/native/Embedding.cpp:
```diff
@@ -15,18 +15,18 @@ Tensor embedding(const Tensor & weight, const Tensor & indices,
   // TODO: use tensor.index() after improving perf
   if (indices.dim() == 1) {
-    return weight.index_select(0, indices);
+    return weight.index_select(0, indices.to(kLong));
```
haha, maybe we should make index_select to be int32 too, but that can wait till another day :)

dzhulgakov commented on aten/src/ATen/native/Embedding.cpp:
```diff
@@ -68,65 +68,63 @@ Tensor embedding_sparse_backward(
   auto index = indices.reshape({1, -1});
   auto values = grad.reshape({-1, num_features});
-  return at::_sparse_coo_tensor_unsafe(index, values, weight_size);
+  return at::_sparse_coo_tensor_unsafe(index.to(kLong), values, weight_size);
```
hm, so we're doing still int64 always in the backwards? are there concerns with doing int32? sparse tensor should support it.

Or the worry is that we'd need to update optimizers too?

ngimel replied:
It doesn't matter, indices of sparse tensor will be converted to Long anyway.

qizzzh replied:
I tried to use int32 initially and found that SparseTensor or at least some SparseTensor utilities assume int64 indices - https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/SparseTensorImpl.cpp.html#90

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -32,11 +32,16 @@ namespace native {
-  offset2bag = offset2bag.cumsum(0);     // offset2bag = [0 0 1 1 2]
+static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
+  AT_DISPATCH_INDEX_TYPES(offsets.scalar_type(), "make_offset2bag", [&] () {
+    auto index_data = offsets.contiguous().data_ptr<index_t>();
```
offsets.contiguous() is a temporary that will be deleted, you should have a local var for it

dzhulgakov commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -32,11 +32,16 @@ namespace native {
+    }
+  });
+  offset2bag[0] -= 1;
+  offset2bag = offset2bag.cumsum(0, offset2bag.scalar_type());
```
nit: since you're manually writing this code - you could directly fold cumsum calculation above - the offsets are always sorted

dzhulgakov reviewed (commented):
Epic, thanks a lot for working on it!

ailzhang commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -32,11 +32,16 @@ namespace native {
-  offset2bag[0] -= 1;                     // offset2bag = [0 0 1 0 1]
-  offset2bag = offset2bag.cumsum(0);     // offset2bag = [0 0 1 1 2]
+static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
+  AT_DISPATCH_INDEX_TYPES(offsets.scalar_type(), "make_offset2bag", [&] () {
```
@qizzzh The reason that XLA test failed is here data storage is directly accessed. From native_functions.yaml _embedding_bag_backward (which calls make_offset2bag) should be device generic (cannot assume storage exists). There're two ways to fix:
1. Use at:: functions instead of directly accessing storage in make_offset2bag
2. Move the old _embedding_bag_backward implementation to Math key and add dispatch section for it in native_functions.yaml. The key here is after this change, your kernel no longer works for all backends. More details can found https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#choosing-the-right-dispatch-keyword
Either solution works for me, the first one might be a little less work I think.

qizzzh replied:
Thanks for the explanation. I guess I'll just make index_add support int32 indices so we keep the original code. index_add is pretty self-contained and I hope it would be fine.

JackCaoG commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -454,7 +460,7 @@ Tensor & index_select_out_cpu_(Tensor & result, const Tensor & self, int64_t dim
   auto numel = index.numel();
   TORCH_CHECK_INDEX(index.dim() <= 1, "index_select(): Index is supposed to be a vector");
-  TORCH_CHECK(index.scalar_type() == ScalarType::Long, "index_select(): Expected dtype int64 for index");
+  TORCH_CHECK(index.scalar_type() == ScalarType::Long || index.scalar_type() == ScalarType::Int, "index_select(): Expected dtype int64 for index");
```
Update the string to include int32 as well?

ngimel commented on torch/_torch_docs.py:
```diff
@@ -3335,7 +3335,7 @@ def merge_dicts(*dicts):
     {input}
     dim (int): the dimension in which we index
-    index (LongTensor): the 1-D tensor containing the indices to index
+    index (IntTensor or LongTensor): the 1-D tensor containing the indices to index
```
I don't think IntTensor is supported on cuda?

qizzzh replied:
index_add and index_select on cuda are enabled in this PR. Github folds long files. Need to avoid adding more to this PR:)

ngimel reviewed (approved):
This looks good. Can you please file an issue to enable int support for index_add and index_select on cuda?

qizzzh commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -434,12 +437,15 @@ Tensor& index_add_cpu_(Tensor & self, int64_t dim, const Tensor & index, const T
-        scalar_t *self_ip = self_ptr + self_i * self_stride;
-        *self_ip += *(source_ptr + i * source_stride);
-      }
+      AT_DISPATCH_INDEX_TYPES(index.scalar_type(), "index_add_cpu_", [&] () {
```
I'm hitting some windows build failures here - https://app.circleci.com/pipelines/github/pytorch/pytorch/233453/workflows/e2a13402-bc72-474f-89a7-6c1f0132cdd0/jobs/8597430/parallel-runs/0/steps/0-104

Seems variables are not captured correctly in the nested lambda. Also found some asks about similar issues online - https://developercommunity.visualstudio.com/content/problem/1150322/nested-c20-template-lambdas-two-issues-with-captur.html

Hi @peterjc123 ! @ngimel suggested that you might have solution/workaround to this issue. Would you mind taking a look? Thank you in advance!

qizzzh replied:
Explicitly capturing variables could work around it.

Commit:
Support int32 indices and offsets in nn.EmbeddingBag (#46758)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46758

It's in general helpful to support int32 indices and offsets, especially when such tensors are large and need to be transferred to accelerator backends. Since it may not be very useful to support the combination of int32 indices and int64 offsets, here we enforce that these two must have the same type.

Test Plan: unit tests

Differential Revision: D24470808

fbshipit-source-id: c468b9678b10ef937c645a14c84e3094b8a1c7ba
